### ISSUE TITLE
Please considersing correctly locate image without the same size

### ISSUE THREAD

**Issue by fengnex** [rainmanwy/robotframework-SikuliLibrary#129]
The is one CRITICAL issue, which I think has a severe negative impact on the use of the library:
Let's say, you want to find a small image from a big image, and there is really the same image in the big one when checking both through human's eyes, but the part in the big image is actually bigger or smaller than the small image.

Now the library really cannot find the specified image in the above situation in the current version, so people have to repeatedly and endlessly crop the same images whenever using the library in different devices. As a consequence, many people stop using the library just because of the problem and I believe that we can see obvious soaring in the library's use after addressing the issue.

However, according to my knowledge on OpenCV, its method named matchTemplate  can be used to solve the problem by resizing one of both images first and I have noticed that SikuliX1 had also made some progress in the aspect such as the following:
https://github.com/RaiMan/SikuliX1/commit/2192b435d0c7aaa00bf0352b5f58cbc206bbd285

So, would you mind fix the problem as soon as possible?@rainmanwy

**Comment by rainmanwy** [comment]
@fengnex , thank you for your advice. 
So you mean add a option(for example resizeFactor) to resize the cropped images, am i right?

**Comment by rainmanwy** [comment]
And could you attach the example images here?

**Comment by RaiMan** [comment]
@fengnex 
As long, as there are systematic changes to how captured images are rendered on other systems, then SikuliX should be able to apply these changes to the images before searching. But this is always something the script user has to address either by using a feature of SikuliX or some available options.
Resize is one prominent example for such a systematic change and is the only option implemented in SikuliX currently, globally (Settings.AlwaysResize) as well as for a single image using Pattern().resize(). But this only works, if the different rendering can be simulated by a simple image resize.

@rainmanwy 
Hope it helps.

**Comment by fengnex** [comment]
@rainmanwy 
Let me give you an example in python where OpenCV-python has been used:
`
template = cv2.imread("template")
template = cv2.cvtColor(template, cv2.COLOR_BGR2GRAY)
template = cv2.Canny(template, 50, 200)
(tH, tW) = template.shape[:2]

for each image:
	gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
	found = None
	confidence = 0.
	
	for scale in np.linspace(0.2, 1.5, 20)[::-1]:		
		resized = imutils.resize(gray, width = int(gray.shape[1] * scale))
		r = gray.shape[1] / float(resized.shape[1])		
		if resized.shape[0] < tH or resized.shape[1] < tW:
			break

		edged = cv2.Canny(resized, 50, 200)
		result = cv2.matchTemplate(edged, template, cv2.TM_CCOEFF)
		(_, maxVal, _, maxLoc) = cv2.minMaxLoc(result)
`
In other words, by resizing and invoking cv2.matchTemplate, we can try to find the maximum matching result, and this value can help us find the expected value in many situations. What I recommend is that a similar way in java should be developed too.

**Comment by fengnex** [comment]
`template = cv2.imread(args["template"])
template = cv2.cvtColor(template, cv2.COLOR_BGR2GRAY)
template = cv2.Canny(template, 50, 200)
(tH, tW) = template.shape[:2]
cv2.imshow("Template", template)

for imagePath in glob.glob(path + "/*.png"):
	image = cv2.imread(imagePath)
	gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)	
	for scale in np.linspace(0.2, 1.5, 20)[::-1]:	
		resized = imutils.resize(gray, width = int(gray.shape[1] * scale))
		r = gray.shape[1] / float(resized.shape[1])
		if resized.shape[0] < tH or resized.shape[1] < tW:
			break
		
		edged = cv2.Canny(resized, 50, 200)
		result = cv2.matchTemplate(edged, template, cv2.TM_CCOEFF)		
		(_, maxVal, _, maxLoc) = cv2.minMaxLoc(result)`

**Comment by RaiMan** [comment]
@fengnex @rainmanwy 
as mentioned earlier: If your problem can be solved by systematically resizing a given image to be searched for, then this is already available in SikuliX.
> Resize is one prominent example for such a systematic change and is the only option implemented in SikuliX currently, globally (Settings.AlwaysResize) as well as for a single image using Pattern().resize()

**RaiMan from SikuliX**

**Comment by fengnex** [comment]
Maybe these measures matter, such as converting the image into a gray one (eliminate color difference）, invoking the method Canny and then call matchTemplate.

**Comment by RaiMan** [comment]
@fengnex 
What you are doing is to create grayscale images (template and image), that are only containing edges.

This is indeed an interesting approach, but might also have problematic cases (more than one match with the same edges e.g. buttons). In SikuliX we currently use BGR-Mats with matchTemplate. If the basic need (the searched image exists with the same width/height in pixels in the template image) SikuliX works in all cases.

But you are not talking about this "normal" case, you are talking about situations, where the image does NOT exist with the same size in pixels in the template.

As mentioned: SikuliX has a feature to do the needed resize in case. But this feature has to be made available in @rainmanwy's SikuliLibrary

**Comment by RaiMan** [comment]
@rainmanwy 
If it makes sense: I can help to add the resize feature in a sensible way.

**Comment by rainmanwy** [comment]
@RaiMan @fengnex . Thank you very much!
@RaiMan , it will be helpful to get your PR.

**Comment by fengnex** [comment]
@rainmanwy 
You are welcome and hope to see the new feature will be coming soon!

**Comment by fengnex** [comment]
Is there any message on when we can use the new feature?

**Comment by RaiMan** [comment]
Uuuups - sorry.
I missed that.
Anyway I am a bit overloaded with other private priorities currently.
I will try my very best, but it might still get February.

**Comment by fengnex** [comment]
Don't know when we can find that from robotframework-SikuliLibrary then.

**Comment by fengnex** [comment]
Is there any information on when we can use the new feature discussed here now?
